# AlpesHearth

## Caso de estudio AlpesHearth. 

El jupyter notebook aqui presente se basa en el caso de estudio AlpesHearth, el cual tiene como objetivos:
- Aplicar técnicas de regresión para construir un modelo predictivo que permita estimar el riesgo cardiovascular de una persona siguiendo el ciclo de machine learning.
- Determinar los principales factores de riesgo cardiovascular con base en los datos.
- Aplicar y comprender un modelo de regresión lineal.
- Reconocer posibles sesgos del modelo de aprendizaje de máquina.
- Comunicar de forma clara y sintética los resultados obtenidos.

Se utilizaron las siguientes librerias de Python para el procesamiento y analisis de datos: 
- Pandas
- Scikit-Learn
- Matplotlib, Seaborn

#### Importar las librerias

In [41]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns 
from IPython.display import display

In [4]:
from importlib.metadata import version
print(f"Versión de Pandas: {version('pandas')}")
print(f"Versión de Numpy: {version('numpy')}")
print(f"Versión de Matplotlib: {version('matplotlib')}")
print(f"Versión de Seaborn: {version('seaborn')}")

Versión de Pandas: 2.3.3
Versión de Numpy: 2.2.6
Versión de Matplotlib: 3.10.8
Versión de Seaborn: 0.13.2


### Exploración de los datos

In [46]:
datos_entrenamiento = pd.read_csv('./Datos/Datos Lab 1.csv')

In [47]:
data = datos_entrenamiento.copy()

In [48]:
data.head()

Patient ID    Date of Service Sex   Age  Weight (kg)  Height (m)     BMI  \
0   isDx5313  November 08, 2023   M  44.0      114.300       1.720  38.600   
1   LHCK2961         20/03/2024   F  57.0       92.923       1.842  33.116   
2   WjVn1699         2021-05-27   F   NaN       73.400       1.650  27.000   
3   dCDO1109     April 18, 2022   F  35.0      113.300       1.780  35.800   
4   pnpE1080         01/11/2024   F  48.0      102.200       1.750  33.400   

   Abdominal Circumference (cm) Blood Pressure (mmHg)  \
0                       100.000                112/83   
1                       106.315                101/91   
2                        78.100                 90/74   
3                        79.600                 92/89   
4                       106.700                121/68   

   Total Cholesterol (mg/dL)  ...  Physical Activity Level  \
0                      228.0  ...                     High   
1                      158.0  ...                     High   
2                      135.0  ...                     High   
3                      158.0  ...                 Moderate   
4                      207.0  ...                      Low   

   Family History of CVD Height (cm) Waist-to-Height Ratio Systolic BP  \
0                      N     172.000                 0.581       112.0   
1                      Y     184.172                 0.577       101.0   
2                      N     165.000                 0.473        90.0   
3                      Y     178.000                 0.447        92.0   
4                      Y     175.000                 0.610       121.0   

  Diastolic BP  Blood Pressure Category  Estimated LDL (mg/dL)  \
0         83.0     Hypertension Stage 1                  121.0   
1         91.0     Hypertension Stage 2                   57.0   
2         74.0                   Normal                   45.0   
3         89.0     Hypertension Stage 1                   94.0   
4         68.0                 Elevated                  128.0   

   CVD Risk Score  CVD Risk Level  
0          19.880            HIGH  
1          16.833    INTERMEDIARY  
2          12.600             LOW  
3          14.920            HIGH  
4          18.870            HIGH  

[5 rows x 24 columns]

Aqui podemos ver como hay columnas de datos repetidos y/o muy parecidos. Por ejemplo, Height (m) y Height (cm), CVD Risk Score y CVD Risk Level, Waist-to-Height Ratio Height y Abdominal Circumference, BMI Weight y Height etc.

A la hora de realizar el analisis estadistico, sera necesario ver cuales de estas variables son redundantes y no nos aportan al modelo. Asimismo, hay que revisar para casos tales como las dos estaturas, si coinciden. De no coincidir, habra que realizar un promedio o utilizar alguna otra tecnica.

In [49]:
display(data.sample(5))

Patient ID    Date of Service Sex   Age  Weight (kg)  Height (m)     BMI  \
1136   mxFg5746         16/11/2024   M  47.0      118.100         NaN  35.700   
942    vLaI6068         22/11/2021   F  72.0       96.579       1.864  35.783   
1045   DYFB4591  February 16, 2025   M   NaN       52.800       1.800  16.300   
319    ZGGT4348         2021-03-13   F  58.0       97.126       1.720     NaN   
134    viiE0140         2020-09-09   F  41.0       82.337       1.727  18.828   

      Abdominal Circumference (cm) Blood Pressure (mmHg)  \
1136                        93.400                130/87   
942                         93.996                114/91   
1045                        77.300                113/96   
319                         81.804                 94/63   
134                         72.053                154/70   

      Total Cholesterol (mg/dL)  ...  Physical Activity Level  \
1136                      151.0  ...                      Low   
942                       282.0  ...                 Moderate   
1045                      167.0  ...                      Low   
319                       159.0  ...                     High   
134                       207.0  ...                 Moderate   

      Family History of CVD Height (cm) Waist-to-Height Ratio Systolic BP  \
1136                      Y     182.000                 0.513       130.0   
942                       Y     186.374                 0.504       114.0   
1045                      N     180.000                 0.429       113.0   
319                       Y     172.019                 0.476        94.0   
134                       Y     172.717                 0.417       154.0   

     Diastolic BP  Blood Pressure Category  Estimated LDL (mg/dL)  \
1136         87.0     Hypertension Stage 1                   82.0   
942          91.0     Hypertension Stage 2                  193.0   
1045         96.0     Hypertension Stage 2                  105.0   
319          63.0                   Normal                   40.0   
134          70.0     Hypertension Stage 2                  124.0   

      CVD Risk Score  CVD Risk Level  
1136          18.660            HIGH  
942           18.497             LOW  
1045          14.250    INTERMEDIARY  
319           13.047    INTERMEDIARY  
134           27.267            HIGH  

[5 rows x 24 columns]

In [50]:
data.shape

(1639, 24)

Esto nos dice que el conjunto de datos esta compuesto por 1.639 registros (filas) y 24 variables (columnas). 

In [ ]:
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1639 entries, 0 to 1638
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Patient ID                    1639 non-null   object 
 1   Date of Service               1639 non-null   object 
 2   Sex                           1639 non-null   object 
 3   Age                           1571 non-null   float64
 4   Weight (kg)                   1566 non-null   float64
 5   Height (m)                    1578 non-null   float64
 6   BMI                           1586 non-null   float64
 7   Abdominal Circumference (cm)  1578 non-null   float64
 8   Blood Pressure (mmHg)         1639 non-null   object 
 9   Total Cholesterol (mg/dL)     1571 non-null   float64
 10  HDL (mg/dL)                   1557 non-null   float64
 11  Fasting Blood Sugar (mg/dL)   1585 non-null   float64
 12  Smoking Status                1639 non-null   object 
 13  Dia

Arriba podemos ver el detalle de cada columna, como el nombre, el número de valores no nulos (en este caso al tener 1.639, indica que no se tienen valores faltantes) y el tipo de los datos de cada columna, al igual que el número de columnas con cada tipo.

Es interesante ver que las columnas que no son numéricas (int o float) son `Patient ID`, `Date of Service`, `Sex`, `Blood Pressure (mmHg)`, `Smoking Status`, `Diabetes Status`, `Physical Activity Level`, `Family History of CVD`, `Blood Pressure Category`, y `CVD Risk Level`, las cuales están en formato object, esto es algo importante y que debemos tener en cuenta.

Un insumo imprescindible para entender un conjunto de datos es el diccionario, ya que este nos permite conocer el significado de cada variable y sus rangos válidos.  

Al mismo tiempo, al revisar los valores non-null de cada una de las columnas, podemos ver que hay muchos datos NULL dentro de la base de datos. Por lo que se debera realizar algo al respecto.

In [56]:
cols_con_null = []

for col in data.columns:
    if data[col].count() != len(data):
        cols_con_null.append(col)

cols_con_null

['Age',
 'Weight (kg)',
 'Height (m)',
 'BMI',
 'Abdominal Circumference (cm)',
 'Total Cholesterol (mg/dL)',
 'HDL (mg/dL)',
 'Fasting Blood Sugar (mg/dL)',
 'Height (cm)',
 'Waist-to-Height Ratio',
 'Systolic BP',
 'Diastolic BP',
 'Estimated LDL (mg/dL)',
 'CVD Risk Score']

In [52]:
diccionario = pd.read_excel('./Datos/DiccPacientes.xlsx')
pd.set_option('display.max_colwidth', None)
diccionario

Nombre Columna  Tipo de dato  \
0                     Patient ID       String   
1                Date of Service         Date   
2                            Sex       String   
3                            Age      Integer   
4                    Weight (kg)        Float   
5                     Height (m)        Float   
6                            BMI        Float   
7   Abdominal Circumference (cm)        Float   
8          Blood Pressure (mmHg)       String   
9      Total Cholesterol (mg/dL)        Float   
10                   HDL (mg/dL)        Float   
11   Fasting Blood Sugar (mg/dL)        Float   
12                Smoking Status      Boolean   
13               Diabetes Status      Boolean   
14       Physical Activity Level       String   
15         Family History of CVD      Boolean   
16                   Height (cm)        Float   
17         Waist-to-Height Ratio        FLoat   
18                   Systolic BP      Integer   
19                  Diastolic BP      Integer   
20       Blood Pressure Category       String   
21         Estimated LDL (mg/dL)        Float   
22                CVD Risk Score        Float   
23                CVD Risk Level      String    

                                                                                              Comentarios  
0                                                                              Identificador del paciente  
1                                                                                    Fecha de la atención  
2                                                                              Sexo (Femenino, Masculino)  
3                                                                                                    Edad  
4                                                                                                    Peso  
5                                                                                                  Altura  
6                                                                                 Índice de masa corporal  
7                                                                                Circunferencia abdominal  
8             Presión sanguínea, de la forma "<Presión arterial sistólica>/<Presión arterial diastólica>"  
9                                                                                        Colesterol total  
10                                                 Lipoproteína de alta densidad en el torrente sanguíneo  
11                                                                  Nivel de azucar en la sangre en ayuno  
12                                                          Booleano que indica si el paciente es fumador  
13                                                        Booleano que indica si el paciente es diabético  
14                                                       Nivel de actividad física (Alto, Moderado, Bajo)  
15             Booleano que indica si en la familia han habido antecedentes de enfermedad cardio vascular  
16                                                                                               Estatura  
17                                             Cociente entre la circumferencia de la cintura y la altura  
18                                                                             Presión arterial sistólica  
19                                                                            Presión arterial diastólica  
20  Categoría de presión sanguínea (Normal, Elevada, Hipertensión Etapa 1, Elevada, Hipertensión Etapa 2)  
21                                                 Lipoproteína de baja densidad en el torrente sanguíneo  
22                                                                                  Riesgo cardiovascular  
23                                                 Nivel de riesgo cadiovascular (Alto, Intermedio, Bajo)

In [53]:
data.describe()

Age  Weight (kg)   Height (m)          BMI  \
count  1571.000000  1566.000000  1578.000000  1586.000000   
mean     46.803186    85.666006     1.757439    28.424744   
std      13.039479    21.712504     0.118012     7.309275   
min       6.134000    13.261000     1.371000     4.317000   
25%      37.000000    67.100000     1.666500    22.600000   
50%      46.000000    86.314000     1.760000    28.000000   
75%      55.000000   104.801500     1.850000    33.963000   
max      89.420000   158.523000     2.146000    53.028000   

       Abdominal Circumference (cm)  Total Cholesterol (mg/dL)  HDL (mg/dL)  \
count                   1578.000000                1571.000000  1557.000000   
mean                      91.538861                 199.043673    56.183558   
std                       13.427985                  59.388670    16.721702   
min                       49.542000                  -1.256000     0.008000   
25%                       79.700000                 150.000000    42.000000   
50%                       91.200000                 199.000000    56.000000   
75%                      102.267250                 250.000000    70.000000   
max                      136.336000                 385.679000   110.315000   

       Fasting Blood Sugar (mg/dL)  Height (cm)  Waist-to-Height Ratio  \
count                  1585.000000  1571.000000            1563.000000   
mean                    117.836860   175.770082               0.522440   
std                      32.379634    11.695880               0.085692   
min                      15.306000   136.498000               0.250000   
25%                      92.000000   167.000000               0.453000   
50%                     115.000000   176.000000               0.519000   
75%                     139.000000   185.000000               0.582000   
max                     219.667000   214.394000               0.804000   

       Systolic BP  Diastolic BP  Estimated LDL (mg/dL)  CVD Risk Score  
count  1578.000000   1554.000000            1582.000000     1610.000000  
mean    125.632637     82.887536             113.235896       18.227281  
std      22.577463     15.503625              61.435291       10.767666  
min      49.914000     31.720000             -92.055000      -20.057000  
25%     108.000000     71.000000              62.000000       15.150000  
50%     125.000000     82.000000             112.000000       16.967000  
75%     141.000000     93.000000             159.000000       18.900000  
max     202.711000    134.066000             317.314000      114.980000

Podemos ver como, para Height (m) y Height (cm), la estatura minima y maxima de ambos rubros no coinciden, por lo que habra que realizar algun tipo de operacion. 

Para el peso seria prudente revisar si es posible? Es decir si se podria un individuo de 13kg con 20 anios? 

Un valor de 0.085 de Waist to Height es extremadamente inusual y parece ser inventado. Es posible que no sea un valor correcto. Posiblemente un valor un poco inferior a 0.35 (aunque ya muy bajo) pordia ser posible. 

Un Systolic BP de 22.577 es sumamente improbable y muy posiblemente un error de medicion. 

Un Estimated LDL de -92.05 es fisiologicamente imposible. Por lo que es posible que este valor no sea real. 

Un valor negativo de CVD Risk Score no es posible. Los valores posibles son entre 0 y 100. Por lo que hay errores en los minimos y maximos.

In [ ]:
# Revisar formatos, asegurarse que no hayan representaciones diferentes de los mismos datos...

objetos = ['Patient ID', 'Date of Service', 'Sex', 
           'Blood Pressure (mmHg)', 'Smoking Status', 
           'Diabetes Status', 'Physical Activity Level', 
           'Family History of CVD', 'Blood Pressure Category', 
           'CVD Risk Level'
] 

for i in objetos:
    display(data[i].value_counts())

Patient ID
AhYt1346    3
RlsB8509    3
RwGu5647    3
dJuC5084    3
YLCe2926    3
           ..
Srzz2840    1
nMHG7307    1
VpjT3887    1
CNBz8190    1
qgQI1151    1
Name: count, Length: 1376, dtype: int64

Date of Service
09-20-2023           6
December 05, 2025    5
December 02, 2020    4
01/12/2024           4
08-09-2022           4
                    ..
2025-02-16           1
April 14, 2021       1
07/09/2021           1
12-13-2024           1
22/10/2021           1
Name: count, Length: 1274, dtype: int64

Sex
M    821
F    818
Name: count, dtype: int64

Blood Pressure (mmHg)
124/72     6
127/84     6
121/68     5
129/61     5
119/88     4
          ..
121/66     1
112/117    1
129/86     1
143/100    1
106/83     1
Name: count, Length: 1152, dtype: int64

Smoking Status
Y    850
N    789
Name: count, dtype: int64

Diabetes Status
N    821
Y    818
Name: count, dtype: int64

Physical Activity Level
High        582
Moderate    537
Low         520
Name: count, dtype: int64

Family History of CVD
N    820
Y    819
Name: count, dtype: int64

Blood Pressure Category
Hypertension Stage 2    680
Hypertension Stage 1    527
Normal                  321
Elevated                111
Name: count, dtype: int64

CVD Risk Level
HIGH            793
INTERMEDIARY    616
LOW             230
Name: count, dtype: int64

Aqui podemos ver tanto que hay IDs repetidos, como que hay formatos diferentes para las fechas de servicio, por lo que sera necesario revisar los IDs, y refactorizar las fechas a un formato estandar.

In [55]:
datos_prueba = pd.read_csv(
    './Datos/Datos Test Lab 1.csv',
    sep=';'
)